JNI Reader for Table Iterator

swamirishi opened the pull request:
Adding a Jni layer for table Iterator. Also changing the table iterator's interface preparse internal key so that SSTFileReaderIterator can be reused.

swamirishi committed:
Jni for TableIterator

swamirishi committed:
Merge remote-tracking branch 'origin/main' into HEAD

swamirishi commented:
@jowlyzhang Can you please review this? In ozone we depend on jni for rocksdb apis. I have just added JNI layer for the table_iterator written in your PR.

swamirishi committed:
Merge branch 'main' into JniReaderForTableIterator

adamretter commented on java/src/main/java/org/rocksdb/RawSstFileReaderIterator.java:
```diff
@@ -0,0 +1,86 @@
+import java.util.Map;
+
+/**
+ * <p>An iterator that yields a sequence of internal key/value pairs from a source.
```
Is this Javadoc correct and specific to this iterator type?

swamirishi replied:
yeah

adamretter commented on java/src/main/java/org/rocksdb/RawSstFileReaderIterator.java:
```diff
@@ -0,0 +1,86 @@
+    kTypeColumnFamilyValuePreferredSeqno(0x19),
+    kMaxValue(0x7F);
+    private int value;
+    private static Map<Integer, ValueType> reverseLookupMap = new HashMap<>();
```
This should instead be handled via a type mapping in `portal.h`, see other examples of Enum types in RocksJava e.g. `TickerType.java`

swamirishi replied:
done

adamretter commented on java/rocksjni/raw_sst_file_reader_iterator.cc:
```diff
@@ -0,0 +1,40 @@
+ * Method:    type
+ * Signature: (J)[B
+ */
+jint Java_org_rocksdb_RawSstFileReaderIterator_type(JNIEnv* env, jclass /*jcls*/,
```
This would be better handled via a type mapping in `portal.h`, see other examples of Enum types in RocksJava e.g. `TickerType.java`

swamirishi replied:
done

adamretter commented on java/src/main/java/org/rocksdb/RawSstFileReaderIterator.java:
```diff
@@ -0,0 +1,86 @@
+ *
+ * @see RocksObject
+ */
+ class RawSstFileReaderIterator extends SstFileReaderIterator {
```
Did you intend for this class to be package-private scope? i.e. not usable for anyone outside `org.rocksdb`?

swamirishi replied:
Yeah I missed it, I fixed it.

adamretter commented on java/src/main/java/org/rocksdb/RawSstFileReaderIterator.java:
```diff
@@ -0,0 +1,86 @@
+ */
+ class RawSstFileReaderIterator extends SstFileReaderIterator {
+
+  protected RawSstFileReaderIterator(SstFileReader reader, long nativeHandle) {
```
Please set all arguments and variables to `final` where possible.

swamirishi replied:
done

adamretter commented on java/src/main/java/org/rocksdb/RawSstFileReaderIterator.java:
```diff
@@ -0,0 +1,86 @@
+ */
+ class RawSstFileReaderIterator extends SstFileReaderIterator {
+
+  protected RawSstFileReaderIterator(SstFileReader reader, long nativeHandle) {
```
This class needs to implement some distinct form of `disposeInternal` to avoid leaking memory when it is finished with.

swamirishi replied:
It extends SstFileReaderIterator which clears off the required handle.

adamretter commented on table/sst_file_reader.cc:
```diff
@@ -113,7 +113,7 @@ std::unique_ptr<Iterator> SstFileReader::NewTableIterator() {
     return nullptr;
   }
-  return std::make_unique<TableIterator>(internal_iter);
+  return std::make_unique<TableIterator>(internal_iter, &rep_->options);
```
@ajkr @pdillinger Can you comment on this change in the internals for RocksDB please?

adamretter commented on table/table_iterator.h:
```diff
@@ -48,22 +49,65 @@ class TableIterator : public Iterator {
+  }
   void SeekForPrev(const Slice& target) override {
-    return iter_->SeekForPrev(target);
+      std::string seek_key_buf;
```
@ajkr @pdillinger Can you comment on this change in the internals for RocksDB please?

adamretter reviewed (changes requested):
Thanks for the PR. I have left some feedback on the Java changes which need some amendment - also a Java test needs to be added please for this new API. I will leave @ajkr @pdillinger and the RocksDB team to comment on the internal C++ changes to RocksDB table... as I cannot myself say if they are desirable or not for the RocksDB team.

jowlyzhang commented on table/table_iterator.h:
```diff
@@ -48,22 +49,65 @@ class TableIterator : public Iterator {
   void SeekToFirst() override { return iter_->SeekToFirst(); }
   void SeekToLast() override { return iter_->SeekToLast(); }
-  void Seek(const Slice& target) override { return iter_->Seek(target); }
+  void Seek(const Slice& target) override {
```
I will review this. Thanks!

jowlyzhang commented on table/table_iterator.h:
```diff
@@ -48,22 +49,65 @@ class TableIterator : public Iterator {
   void SeekToFirst() override { return iter_->SeekToFirst(); }
   void SeekToLast() override { return iter_->SeekToLast(); }
-  void Seek(const Slice& target) override { return iter_->Seek(target); }
+  void Seek(const Slice& target) override {
```
The type of key taken as input for `Seek` and `SeekForPrev` should be consistent with the type of key produced by `Iterator::key`. Can you add jni APIs for the util methods `GetInternalKeyForSeek` etc instead?

swamirishi replied:
Can't we change the definition of the TableIterator to seek and get only on the user_key instead? Isn't this a better way to define this class?

jowlyzhang replied:
The name `TableIterator` indicates it's iterating the raw table, and that makes internal key a better format of key for these APIs.

jowlyzhang commented on table/table_iterator.h:
```diff
@@ -48,22 +49,65 @@ class TableIterator : public Iterator {
     return Status::NotSupported("TableIterator does not support GetProperty.");
   }
 
+  uint64_t SequenceNumber() {
```
The public API is `Iterator` in rocksdb/includes/iterator.h, the `TableIterator` is an internal class that is not supposed to be publicly accessible. These `SequenceNumber`, and `type` APIs are not defined for `Iterator`.


We have added `ParseEntry` util methods to parse sequence number and type out of an internal key. You can have you own wrapper iterator class wrapping a RocksDB `Iterator`, and add the parsing logic in to the wrapper iterator.

swamirishi replied:
We need all of these infos separately and storing the parsed entry would be much more optimal. Otherwise for each and every call we would have to parse the key over and over again. All I am doing here is changing the definition of the table iter which would seek over the same user key and return user key when getKey() is called. I have added 2 more functions on top of iterator, I believe that should be fine, since JNI would be called only when TableIterator is initialized.

swamirishi replied:
@jowlyzhang TableIterator itself is a wrapper for InternalIterator why have 2 wrappers? Should we consider making this public?

jowlyzhang replied:
It's worth having 2 wrappers because it will be your own wrapper iterator that you don't need to check into RocksDB codebase, you own those logic and you can modify it however you want. Parsing an entry every time over and over again isn't computationally heavy or performance wise concerning, you can just wrap it in your own wrapper iterator. It's beset for the RocksDB `TableIterator` to just do the most vanilla things of iterating raw entries. Others wishing to add their own iterating constraints can just add on top of it.

swamirishi replied:
@jowlyzhang I have updated the code to add another jni functionality to convert internal Key to user key and vice versa

swamirishi committed:
Address review comments

swamirishi committed:
Merge remote-tracking branch 'swamirishi/JniReaderForTableIterator' into HEAD

swamirishi committed:
Merge remote-tracking branch 'origin/main' into HEAD

swamirishi committed:
Merge remote-tracking branch 'origin/main' into HEAD

swamirishi committed:
Jni to convert userKey to internalKey and vice versa

swamirishi committed:
Merge remote-tracking branch 'origin/main' into JniReaderForTableIterator_wip

swamirishi committed:
Add javadoc

swamirishi committed:
revert jni opt

swamirishi committed:
Merge remote-tracking branch 'origin/main' into JniReaderForTableIterator_wip

swamirishi commented:
@jowlyzhang Can you check this PR?

jowlyzhang commented:
> @jowlyzhang Can you check this PR?

Thanks for making the changes. Since this PR now mostly have java changes, I'll leave it to @adamretter to make the call here.

adamretter commented on java/src/main/java/org/rocksdb/EntryType.java:
```diff
@@ -0,0 +1,44 @@
+  kEntryWideColumnEntity((byte)0x7),
+  kEntryOther((byte)0x8);
+  private final byte value;
+  private static Map<Byte, EntryType> reverseLookupMap = new HashMap<>();
```
Do we need a map here? The collection is small enough to iterate I would have thought?

swamirishi replied:
removed the map and falling back to iterating through the values.

adamretter commented on java/src/main/java/org/rocksdb/EntryType.java:
```diff
@@ -0,0 +1,44 @@
+  private final byte value;
+  private static Map<Byte, EntryType> reverseLookupMap = new HashMap<>();
+
+  EntryType(byte value) {
```
Should be `final` param.

swamirishi replied:
done

adamretter commented on java/rocksjni/parsed_entry_info.cc:
```diff
@@ -0,0 +1,216 @@
+                                                                  jint len) {
+  const std::unique_ptr<char[]> target(new char[len]);
+  if (target == nullptr) {
+    jclass oom_class = env->FindClass("/lang/java/OutOfMemoryError");
```
I don't think this has been tested as the class name is incorrect. Also there are util methods in portal.h for throwing errors that could be used.

swamirishi replied:
Added util method to also do operations on indirect byte array buffers

adamretter commented on java/rocksjni/parsed_entry_info.cc:
```diff
@@ -0,0 +1,216 @@
+                  "Memory allocation failed in RocksDB JNI function");
+    return;
+  }
+  env->GetByteArrayRegion(jtarget, off, len,
```
Needs error checking

swamirishi replied:
done

adamretter commented on java/rocksjni/parsed_entry_info.cc:
```diff
@@ -0,0 +1,216 @@
+  auto slice_size = key_slice.size();
+  jsize copy_size = std::min(static_cast<uint32_t>(slice_size),
+                             static_cast<uint32_t>(jlen));
+  env->SetByteArrayRegion(
```
Needs error checking

swamirishi replied:
done

adamretter commented on java/rocksjni/parsed_entry_info.cc:
```diff
@@ -0,0 +1,216 @@
+    // exception thrown: OutOfMemoryError
+    return nullptr;
+  }
+  env->SetByteArrayRegion(
```
Needs error checking

swamirishi replied:
done

adamretter commented on java/rocksjni/type_util.cc:
```diff
@@ -0,0 +1,311 @@
+      reinterpret_cast<const ROCKSDB_NAMESPACE::Options *>(options_handle);
+  const std::unique_ptr<char[]> target(new char[user_key_len]);
+  if (target == nullptr) {
+    jclass oom_class = env->FindClass("/lang/java/OutOfMemoryError");
```
Same concern about incorrect class name and testing here.

adamretter commented on java/rocksjni/type_util.cc:
```diff
@@ -0,0 +1,311 @@
+                  "Memory allocation failed in RocksDB JNI function");
+    return static_cast<jsize>(0);
+  }
+  env->GetByteArrayRegion(user_key, user_key_off, user_key_len,
```
Needs error checking.

adamretter commented on java/rocksjni/type_util.cc:
```diff
@@ -0,0 +1,311 @@
+  auto slice_size = key_slice.size();
+  jsize copy_size = std::min(static_cast<uint32_t>(slice_size),
+                             static_cast<uint32_t>(int_key_len));
+  env->SetByteArrayRegion(
```
Needs error checking.

swamirishi replied:
done

adamretter commented on java/rocksjni/type_util.cc:
```diff
@@ -0,0 +1,311 @@
+      reinterpret_cast<const ROCKSDB_NAMESPACE::Options *>(options_handle);
+  const std::unique_ptr<char[]> target(new char[user_key_len]);
+  if (target == nullptr) {
+    jclass oom_class = env->FindClass("/lang/java/OutOfMemoryError");
```
Class name and testing?

swamirishi replied:
done

adamretter commented on java/rocksjni/type_util.cc:
```diff
@@ -0,0 +1,311 @@
+                  "Memory allocation failed in RocksDB JNI function");
+    return static_cast<jsize>(0);
+  }
+  env->GetByteArrayRegion(user_key, user_key_off, user_key_len,
```
Needs error checking.

swamirishi replied:
done

adamretter commented on java/rocksjni/type_util.cc:
```diff
@@ -0,0 +1,311 @@
+  auto slice_size = key_slice.size();
+  jsize copy_size = std::min(static_cast<uint32_t>(slice_size),
+                             static_cast<uint32_t>(int_key_len));
+  env->SetByteArrayRegion(
```
Needs error checking.

swamirishi replied:
done

adamretter commented on java/rocksjni/type_util.cc:
```diff
@@ -0,0 +1,311 @@
+    // exception thrown: OutOfMemoryError
+    return nullptr;
+  }
+  env->SetByteArrayRegion(
```
Needs error checking.

swamirishi replied:
done

adamretter reviewed (changes requested):
The issues I think I have identified repeat throughout and should be addressed please

swamirishi committed:
Merge remote-tracking branch 'origin/main' into JniReaderForTableIterator_wip

swamirishi committed:
Fix OutOfMemoryError

swamirishi committed:
Address review commits

swamirishi committed:
Merge remote-tracking branch 'origin/main' into JniReaderForTableIterator_wip

swamirishi commented:
@adamretter Apologies for the late response and not following through addressing review comments on time. I have addressed all the review comments. Can you please take a look at it? I have added a Java test for the same as well and also added required util functions as well in portal.h. FYI these util functions can be used in other JNI classes as well.

swamirishi committed:
Make final

swamirishi committed:
Merge branch 'main' into JniReaderForTableIterator

swamirishi committed:
Merge branch 'main' into JniReaderForTableIterator

swamirishi commented:
@adamretter can you please take a look at this?

swamirishi committed:
Merge branch 'main' into JniReaderForTableIterator

swamirishi committed:
Merge branch 'main' into JniReaderForTableIterator

swamirishi commented:
@adamretter @jowlyzhang Can you please take a look at this patch, whenever you find time?

adamretter commented on java/src/main/java/org/rocksdb/ParsedEntryInfo.java:
```diff
@@ -0,0 +1,88 @@
+    return result;
+  }
+
+  public void parseEntry(Options options, byte[] internalKey) {
```
Please mark parameters as `final`. (and methods/functions below)

swamirishi replied:
done

adamretter commented on java/src/main/java/org/rocksdb/TypeUtil.java:
```diff
@@ -0,0 +1,121 @@
+ * Util class to get internal key for seeking from user key.
+ */
+public class TypeUtil {
+  public static byte[] getInternalKey(byte[] userKey, Options options) {
```
Missing JavaDoc. (and methods below).

swamirishi replied:
done

adamretter commented on java/src/main/java/org/rocksdb/TypeUtil.java:
```diff
@@ -0,0 +1,121 @@
+ * Util class to get internal key for seeking from user key.
+ */
+public class TypeUtil {
+  public static byte[] getInternalKey(byte[] userKey, Options options) {
```
Parameters should be marked `final`. (and methods below)

swamirishi replied:
done

adamretter commented on java/src/test/java/org/rocksdb/SstFileReaderTest.java:
```diff
@@ -219,4 +219,138 @@ public void readSstFile() throws RocksDBException, IOException {
     }
   }
+
+  private void assertInternalKey(final Options options, ParsedEntryInfo parsedEntryInfo,
```
Parameters should be marked final. (and methods below)

swamirishi replied:
done

adamretter commented on java/rocksjni/parsed_entry_info.cc:
```diff
@@ -0,0 +1,188 @@
+      reinterpret_cast<ROCKSDB_NAMESPACE::ParsedEntryInfo *>(handle);
+  jbyte *target = env->GetByteArrayElements(jtarget, nullptr);
+  if (target == nullptr) {
+    jclass oom_class = env->FindClass("java/lang/OutOfMemoryError");
```
There is a `portal.h` utility method for this - `OutOfMemoryErrorJni::ThrowNew(env, msg)` please use it. (and in the methods below)

swamirishi replied:
done

adamretter commented on java/rocksjni/portal.h:
```diff
@@ -2477,6 +2478,35 @@ class JniUtil {
+   * Copies values from jbyte array to slice and performs op on the slice.
+   * for example WriteBatch->Delete
+   *
+   * TODO(AR) could be extended to cover returning ROCKSDB_NAMESPACE::Status
```
Why does this have my initials against the TODO?

adamretter replied:
Also the naming seems odd as this has nothing to do with an Indirect ByteBuffer (instead it appears to work with a byte array)

swamirishi replied:
indirect byte buffer & byte array share the same interface.

adamretter commented on java/rocksjni/portal.h:
```diff
@@ -2477,6 +2478,35 @@ class JniUtil {
+    }
+    const std::unique_ptr<char[]> target(new char[jkey_len]);
+    if (target == nullptr) {
+      jclass oom_class = env->FindClass("java/lang/OutOfMemoryError");
```
Use the utility exception class in portal.h instead please.

swamirishi replied:
done

adamretter commented on java/rocksjni/portal.h:
```diff
@@ -2498,6 +2528,24 @@ class JniUtil {
     return cvalue_len;
   }
+
+  template <class T>
```
Documentation missing for this function.

adamretter replied:
Also the naming seems odd as this has nothing to do with an Indirect ByteBuffer (instead it appears to work with a byte array)

swamirishi replied:
done

adamretter commented on java/rocksjni/portal.h:
```diff
@@ -6396,6 +6444,35 @@ class TxnDBWritePolicyJni {
   }
 };
 
+// The portal class for org.rocksdb.ValueType
```
`ValueType` -> `EntryType`

swamirishi replied:
done

adamretter commented on java/rocksjni/portal.h:
```diff
@@ -6396,6 +6444,35 @@ class TxnDBWritePolicyJni {
+      case ROCKSDB_NAMESPACE::EntryType::kEntryOther:
+        return 0x8;
+      default:
+        return 0x9;
```
There does not appear to be a `0x9` defined in the Java enum, so this will raise an exception. Is that intentional?

swamirishi replied:
yeah

swamirishi replied:
Added an invalid enum on java side to handle this default value.

adamretter commented on java/rocksjni/sst_file_readerjni.cc:
```diff
@@ -71,6 +71,19 @@ jlong Java_org_rocksdb_SstFileReader_newIterator(JNIEnv * /*env*/,
+                                                      jlong jhandle) {
+  auto *sst_file_reader =
+      reinterpret_cast<ROCKSDB_NAMESPACE::SstFileReader *>(jhandle);
+  return GET_CPLUSPLUS_POINTER(sst_file_reader->NewTableIterator().release());
```
I am a bit worried about the `release()` here? What is the ownership model for the iterator in your code please?

swamirishi replied:
It is the same as sstFileReaderIterator. So release should be triggered manually irrespective of sstFileReader class

adamretter commented on java/rocksjni/type_util.cc:
```diff
@@ -0,0 +1,275 @@
+  ROCKSDB_NAMESPACE::Slice key_slice = seek_key_buf;
+  jbyteArray jkey = env->NewByteArray(static_cast<jsize>(key_slice.size()));
+  if (jkey == nullptr) {
+    jclass oom_class = env->FindClass("/java/lang/OutOfMemoryError");
```
Please use the utility exception class in portal.h

adamretter commented on java/rocksjni/type_util.cc:
```diff
@@ -0,0 +1,275 @@
+      target_slice, options->comparator, &seek_key_buf);
+  ROCKSDB_NAMESPACE::Slice key_slice = seek_key_buf;
+  jbyteArray jkey = env->NewByteArray(static_cast<jsize>(key_slice.size()));
+  if (jkey == nullptr) {
```
Please use the utility exception class in portal.h

swamirishi replied:
done

swamirishi committed:
Address review comments

swamirishi commented on java/src/main/java/org/rocksdb/EntryType.java:
```diff
@@ -0,0 +1,35 @@
+  }
+
+  public static EntryType getEntryType(final byte value) {
+    for (EntryType entryType : EntryType.values()) {
```
done

swamirishi commented on java/src/main/java/org/rocksdb/EntryType.java:
```diff
@@ -0,0 +1,35 @@
+        return entryType;
+      }
+    }
+    throw new IllegalArgumentException("Invalid ValueType byte " + value);
```
done

swamirishi commented on java/src/main/java/org/rocksdb/ParsedEntryInfo.java:
```diff
@@ -0,0 +1,88 @@
+    disposeInternalJni(handle);
+  }
+
+  public EntryType getEntryType() {
```
done

swamirishi committed:
Address review comments

swamirishi committed:
Address review comments

swamirishi committed:
Address review comments

swamirishi committed:
Add doc

swamirishi commented:
@adamretter I have addressed all the review comments. Can you please take a look at the patch again?

swamirishi commented:
@adamretter Can you take a look at this?

adamretter commented:
@swamirishi Not all of my comments have been addressed.

swamirishi committed:
Change function name

swamirishi committed:
Merge branch 'main' into JniReaderForTableIterator

swamirishi commented:
> @swamirishi Not all of my comments have been addressed.

@adamretter Yeah sorry I missed one of the comments. I have changed the method name and set the parameters to final. Can you check if this is fine now?

swamirishi commented:
@adamretter can you please look into this I have addressed all the review comments here.

swamirishi committed:
Merge branch 'main' into JniReaderForTableIterator

swamirishi commented:
@adamretter @jowlyzhang Can you please take a look at this patch? I have addressed all the review comments and this has been pending for a while.

adamretter commented:
@swamirishi Apologies, I have been on holiday. I will try and review this shortly...

swamirishi committed:
Merge branch 'main' into JniReaderForTableIterator

swamirishi commented:
@adamretter thanks a lot

swamirishi committed:
Merge branch 'main' into JniReaderForTableIterator

swamirishi commented:
@adamretter Any updates on this?

swamirishi committed:
Merge branch 'main' into JniReaderForTableIterator

swamirishi commented:
@adamretter Can you review this PR or is there anyone else who can review this PR if you are loaded with other stuff

adamretter commented on java/rocksjni/portal.h:
```diff
@@ -2477,6 +2478,33 @@ class JniUtil {
+    }
+    env->GetByteArrayRegion(jkey, jkey_off, jkey_len,
+                            reinterpret_cast<jbyte*>(target.get()));
+    ROCKSDB_NAMESPACE::Slice target_slice(target.get(), jkey_len);
```
Missing error check on the JNI call before this, e.g. `if (env->ExceptionOccurred())`

adamretter commented on java/rocksjni/type_util.cc:
```diff
@@ -0,0 +1,271 @@
+  };
+  ROCKSDB_NAMESPACE::JniUtil::k_op_direct(getInternalKeySeek, env, user_key,
+                                          user_key_off, user_key_len);
+  ROCKSDB_NAMESPACE::Slice key_slice = seek_key_buf;
```
Missing an error check to see if `k_op_direct` raised a Java exception.

adamretter commented on java/rocksjni/type_util.cc:
```diff
@@ -0,0 +1,271 @@
+  };
+  ROCKSDB_NAMESPACE::JniUtil::k_op_direct(getInternalKeySeek, env, user_key,
+                                          user_key_off, user_key_len);
+  ROCKSDB_NAMESPACE::Slice key_slice = seek_key_buf;
```
Missing an error check to see if `k_op_direct` raised a Java exception.

adamretter commented on java/rocksjni/type_util.cc:
```diff
@@ -0,0 +1,271 @@
+                                             &seek_key_buf);
+  };
+  ROCKSDB_NAMESPACE::JniUtil::k_op_direct(getInternalKeySeek, env, user_key,
+                                          user_key_off, user_key_len);
```
Missing an error check to see if `k_op_direct` raised a Java exception.

adamretter commented on java/rocksjni/type_util.cc:
```diff
@@ -0,0 +1,271 @@
+  };
+  ROCKSDB_NAMESPACE::JniUtil::k_op_direct(getInternalKeySeek, env, user_key,
+                                          user_key_off, user_key_len);
+  ROCKSDB_NAMESPACE::Slice key_slice = seek_key_buf;
```
Missing an error check to see if `k_op_direct` raised a Java exception.

swamirishi replied:
I don't think a java exception will be thrown in k_op_direct

swamirishi replied:
Can you point out what specific issue you are pointing towards here?

adamretter commented on java/rocksjni/type_util.cc:
```diff
@@ -0,0 +1,271 @@
+  };
+  ROCKSDB_NAMESPACE::JniUtil::k_op_indirect(getInternalKeySeek, env, user_key,
+                                            user_key_off, user_key_len);
+  ROCKSDB_NAMESPACE::Slice key_slice = seek_key_buf;
```
Missing an error check to see if `k_op_indirect` raised a Java exception.

adamretter commented on java/rocksjni/type_util.cc:
```diff
@@ -0,0 +1,271 @@
+  };
+  ROCKSDB_NAMESPACE::JniUtil::k_op_indirect(getInternalKeySeek, env, user_key,
+                                            user_key_off, user_key_len);
+  ROCKSDB_NAMESPACE::Slice key_slice = seek_key_buf;
```
Missing an error check to see if `k_op_indirect` raised a Java exception.

adamretter commented on java/rocksjni/type_util.cc:
```diff
@@ -0,0 +1,271 @@
+  };
+  ROCKSDB_NAMESPACE::JniUtil::k_op_indirect(getInternalKeySeek, env, user_key,
+                                            user_key_off, user_key_len);
+  ROCKSDB_NAMESPACE::Slice key_slice = seek_key_buf;
```
Missing an error check to see if `k_op_indirect` raised a Java exception.

swamirishi replied:
fixed in k_op_indirect function.

adamretter commented on java/rocksjni/type_util.cc:
```diff
@@ -0,0 +1,271 @@
+  };
+  ROCKSDB_NAMESPACE::JniUtil::k_op_indirect(getInternalKeySeek, env, user_key,
+                                            user_key_off, user_key_len);
+  ROCKSDB_NAMESPACE::Slice key_slice = seek_key_buf;
```
Missing an error check to see if `k_op_indirect` raised a Java exception.

swamirishi replied:
fixed in k_op_indirect function.

adamretter commented on java/rocksjni/parsed_entry_info.cc:
```diff
@@ -0,0 +1,185 @@
+  }
+  ROCKSDB_NAMESPACE::JniUtil::copyToByteArray(
+      env, key_slice, jkey, 0, static_cast<jint>(key_slice.size()));
+  return jkey;
```
Missing an error check to see if `copyToByteArray` raised a Java exception.

swamirishi replied:
added a check in copyBytes function

adamretter commented on java/rocksjni/type_util.cc:
```diff
@@ -0,0 +1,271 @@
+  }
+  ROCKSDB_NAMESPACE::JniUtil::copyToByteArray(
+      env, key_slice, jkey, 0, static_cast<jint>(key_slice.size()));
+  return jkey;
```
Missing an error check to see if `copyToByteArray` raised a Java exception.

adamretter commented on java/rocksjni/type_util.cc:
```diff
@@ -0,0 +1,271 @@
+  }
+  ROCKSDB_NAMESPACE::JniUtil::copyToByteArray(
+      env, key_slice, jkey, 0, static_cast<jint>(key_slice.size()));
+  return jkey;
```
Missing an error check to see if `copyToByteArray` raised a Java exception.

adamretter reviewed (changes requested):
Still some missing error handling that needs to be implemented.

adamretter commented:
@swamirishi Are you still working on this?

swamirishi commented:
> @swamirishi Are you still working on this?
@adamretter Apologies for not addressing the issues. I have been busy working on other stuff. But I am working on it now and would be able to address all the review comments by EOD.

swamirishi committed:
Merge remote-tracking branch 'origin/main' into HEAD

swamirishi committed:
Address review comments for error handling

swamirishi commented on java/rocksjni/parsed_entry_info.cc:
```diff
@@ -0,0 +1,185 @@
+      reinterpret_cast<const ROCKSDB_NAMESPACE::Options *>(options_handle);
+  auto *parsed_entry_info =
+      reinterpret_cast<ROCKSDB_NAMESPACE::ParsedEntryInfo *>(handle);
+  jbyte *target = env->GetByteArrayElements(jtarget, nullptr);
```
Yeah you are right thanks for pointing out the issue

swamirishi commented on java/rocksjni/portal.h:
```diff
@@ -2498,6 +2526,26 @@ class JniUtil {
+  /* Helper for copying value in source into a byte array.
+   */
+  template <class T>
+  static jint copyToByteArray(JNIEnv* env, T& source, jbyteArray jtarget,
```
done, renamed the function copyBytes. I don't think we can really reuse any of the existing function to reuse. I have called the existing copyBytes to create a new byte array instead of calling the copyToByteArray function

swamirishi commented on java/rocksjni/portal.h:
```diff
@@ -2498,6 +2526,26 @@ class JniUtil {
+
+    const jint cvalue_len = static_cast<jint>(source.size());
+    const jint length = std::min(jtarget_len, cvalue_len);
+    env->SetByteArrayRegion(
```
done

swamirishi committed:
Add invalid enum entryType

swamirishi commented:
@adamretter I have addressed most of the review comments to the best of my knowledge. I had a couple of doubts for few review comments and have left comments inline.

swamirishi committed:
Merge branch 'main' into JniReaderForTableIterator

swamirishi commented:
@adamretter Can this be merged?

swamirishi committed:
Merge branch 'main' into JniReaderForTableIterator

swamirishi committed:
Merge branch 'main' into JniReaderForTableIterator

swamirishi commented:
@adamretter Do you think this can be merged?

adamretter commented:
@swamirishi I have asked my colleague here @alanpaxton to take a look at this for you. He will come back to you shortly...

alanpaxton commented on java/rocksjni/parsed_entry_info.cc:
```diff
@@ -0,0 +1,191 @@
+    ROCKSDB_NAMESPACE::ParseEntry(target_slice, options->comparator,
+                                  parsed_entry_info);
+  };
+  ROCKSDB_NAMESPACE::JniUtil::k_op_indirect(parse, env, jtarget, joff, jlen);
```
The slice data constructed by `k_op_indirect` has a lifetime which ends with the `k_op_indirect` call (`std::unique_ptr`), so the `parsed_entry_info` cannot be guaranteed to persist when it is returned. I wonder if it would be easier to make the `ParsedEntryInfo` a complete Java object, with members for key , type, sequence number, and fully instantiate this using JNI from a temporary C++ ParsedEntryInfo.

I also don't like the `copied_user_key` flag in `ParsedEntryInfo`, and I think if you refactored in the way I suggest, it would allow you to remove `copied_user_key`.

alanpaxton commented on java/src/test/java/org/rocksdb/SstFileReaderTest.java:
```diff
@@ -219,4 +219,138 @@ public void readSstFile() throws RocksDBException, IOException {
     }
   }
+
+  private void assertInternalKey(final Options options, final ParsedEntryInfo parsedEntryInfo,
```
Some of the `parseEntry()` variants are not being tested. Please ensure in particular that all the JNI methods you write get exercised.

alanpaxton reviewed (commented):
I have pointed out a few issues around memory management and lifetime, as well as some coding style/encapsulation issues. Happy to discuss further if that will be of assistance.

swamirishi committed:
Merge remote-tracking branch 'origin/main' into HEAD

swamirishi committed:
Address review comments to fix JNI throw exceptions and fixed junit tests to cover all combinations

swamirishi committed:
Add functionality to add range iteration on table iterator

swamirishi committed:
Fix Memory leak in test

swamirishi committed:
Make it Abstract Slice

swamirishi committed:
Fix NPE
